**Hello, welcome to the preprocessing notebook of our data science project: predicting successful startup!**

We will try to include as much details as possible with markdowns :)

# Data Import

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data = pd.read_csv('/Users/danyhersco/data_startup_final.csv')

# Data Exploration

In [3]:
data.shape

(314486, 19)

We extracted the dataset from DBeaver, after running an SQL query. As a result of multiple `LEFT JOIN` to include `industry` and `technology` features, many startups have duplicates.

Consequently, let's check for the number of unique values for each feature of our dataframe.

In [4]:
data.describe(include='all')

name                   website  \
count                314486                    311288   
unique                55665                     54614   
top     Arthur Intelligence  https://www.goarthur.ai/   
freq                    216                       216   
mean                    NaN                       NaN   
std                     NaN                       NaN   
min                     NaN                       NaN   
25%                     NaN                       NaN   
50%                     NaN                       NaN   
75%                     NaN                       NaN   
max                     NaN                       NaN   

                    short_description ipo_status  founded_on went_public_on  \
count                          314483     314486      314486           5383   
unique                          56142          3        1083            507   
top     Limited information available    private  2016-01-01     2021-11-11   
freq                              597     309434       67496            114   
mean                              NaN        NaN         NaN            NaN   
std                               NaN        NaN         NaN            NaN   
min                               NaN        NaN         NaN            NaN   
25%                               NaN        NaN         NaN            NaN   
50%                               NaN        NaN         NaN            NaN   
75%                               NaN        NaN         NaN            NaN   
max                               NaN        NaN         NaN            NaN   

         exited_on  num_funding_rounds last_equity_funding_type  \
count        22095       314486.000000                   288993   
unique        1370                 NaN                       18   
top     2021-02-19                 NaN                     seed   
freq           146                 NaN                   107532   
mean           NaN            3.414276                      NaN   
std            NaN            2.421491                      NaN   
min            NaN            1.000000                      NaN   
25%            NaN            2.000000                      NaN   
50%            NaN            3.000000                      NaN   
75%            NaN            5.000000                      NaN   
max            NaN           22.000000                      NaN   

                                last_equity_funding_total last_funding_at  \
count                                              236615          314486   
unique                                              15347            2905   
top     {"currency":"USD","amount":100000000,"amountUS...      2019-01-01   
freq                                                 4782            2892   
mean                                                  NaN             NaN   
std                                                   NaN             NaN   
min                                                   NaN             NaN   
25%                                                   NaN             NaN   
50%                                                   NaN             NaN   
75%                                                   NaN             NaN   
max                                                   NaN             NaN   

       headquartersCountry headquartersRegion  employeeCount industry_name  \
count               313337             313294  302219.000000        305345   
unique                 162               1392            NaN            40   
top                     US         California            NaN   Health Care   
freq                138635              49880            NaN         29388   
mean                   NaN                NaN      61.950658           NaN   
std                    NaN                NaN     443.224729           NaN   
min                    NaN                NaN       0.000000           NaN   
25%                    NaN         

We have **55665** unique startups in our dataset. The next step will be to remove all duplicates while treating the non-duplicated columns.

Here is how our dataset look like:

In [5]:
data.head()

name                website  \
0  Cardiomo Care, Inc.  https://cardiomo.com/   
1  Cardiomo Care, Inc.  https://cardiomo.com/   
2  Cardiomo Care, Inc.  https://cardiomo.com/   
3  Cardiomo Care, Inc.  https://cardiomo.com/   
4  Cardiomo Care, Inc.  https://cardiomo.com/   

                                   short_description ipo_status  founded_on  \
0  AI-based Remote Patient Monitoring solution to...    private  2016-01-27   
1  AI-based Remote Patient Monitoring solution to...    private  2016-01-27   
2  AI-based Remote Patient Monitoring solution to...    private  2016-01-27   
3  AI-based Remote Patient Monitoring solution to...    private  2016-01-27   
4  AI-based Remote Patient Monitoring solution to...    private  2016-01-27   

  went_public_on exited_on  num_funding_rounds last_equity_funding_type  \
0            NaN       NaN                  10                     seed   
1            NaN       NaN                  10                     seed   
2            NaN       NaN                  10                     seed   
3            NaN       NaN                  10                     seed   
4            NaN       NaN                  10                     seed   

                           last_equity_funding_total last_funding_at  \
0  {"currency":"USD","amount":10000000,"amountUSD...      2020-08-01   
1  {"currency":"USD","amount":10000000,"amountUSD...      2020-08-01   
2  {"currency":"USD","amount":10000000,"amountUSD...      2020-08-01   
3  {"currency":"USD","amount":10000000,"amountUSD...      2020-08-01   
4  {"currency":"USD","amount":10000000,"amountUSD...      2020-08-01   

  headquartersCountry headquartersRegion  employeeCount industry_name  \
0                  US           New York           31.0   Health Care   
1                  US           New York           31.0   Health Care   
2                  US           New York           31.0   Health Care   
3                  US           New York           31.0   Health Care   
4                  US           New York           31.0   Health Care   

  technology_name              announcedOn     stage  \
0        Software  2019-04-15 00:00:00.000   unknown   
1        Software  2020-08-01 00:00:00.000      seed   
2        Software  2017-04-01 00:00:00.000   unknown   
3        Software  2016-08-09 00:00:00.000  pre_seed   
4        Software  2019-02-01 00:00:00.000  pre_seed   

                                         moneyRaised  
0  {"amount": 5652200, "currency": "USD", "amount...  
1  {"amount": 10000000, "currency": "USD", "amoun...  
2  {"amount": 2500000, "currency": "USD", "amount...  
3  {"amount": 13000000, "currency": "USD", "amoun...  
4  {"amount": 12000000, "currency": "USD", "amoun...

# Remove the duplicates through encoding

This problem will be challenging as there is not really fully duplicated rows. As we said above, it is just a result of the `LEFT JOIN` that multiplicated the rows for one company (several industries, several technologies, and several funding rounds). We have to one hot encode the latter features, which is what we will be doing in this section.

At the end of this section, we should reach a shape of `(55665, >18)`.

In [3]:
from useful.variables import industries, technologies

## `industry`